### PR TITLE
6 fix model saving

### PR DESCRIPTION
**Pull request by tanay-gangey**
Added code for model saving

### CODE REVIEW

File: src/runners/sac.py [hunk 1]
```diff
@@ -157,40 +178,41 @@ def eval(self):
...
             try:
                 # Try to save Safety Actor-Critic, if present
                 torch.save(self.safety_actor_critic.state_dict(), path_name)
             except:
                 pass
 
-        elif self.cfg["save_freq"] > 0 and (n_eps + 1 % self.cfg["save_freq"] == 0):
-            path_name = f"{self.cfg['model_save_path']}/{self.cfg['experiment_name']}_episode_{n_eps}.statedict"
+        elif self.runner_config["save_every_nth_episode"] > 0 and (n_eps + 1 % self.runner_config["save_every_nth_episode"] == 0):
+            path_name = f"{self.runner_config['model_save_dir']}/{self.exp_config['experiment_name']}_episode_{n_eps}.statedict"
```
Comment (kchian) [inline src/runners/sac.py]: Should be `{model_save_dir}/{experiment_name}/{experiment_name}_{episode}_{n_eps}.statedict`

**Review by kchian — DISMISSED** [review]
do a `make black` too

**Comment by tanay-gangey** [comment]
I ran make black and on my terminal it says all 39 files unchanged but the automatic lint check fails for some reason
image

**Comment by kchian** [comment]
> I ran make black and on my terminal it says all 39 files unchanged but the automatic lint check fails for some reason image

weirdness. I fixed it.

File: src/runners/sac.py [hunk 2]
```diff
@@ -85,22 +94,33 @@ def run(self):
...
+            # Save every N episodes or when the current episode return is better than the best return
+            # Following the logic of now deprecated checkpoint_model
+            if ep_number % self.runner_config["save_every_nth_episode"] == 0:
+                save_path = f"{self.runner_config['model_save_dir']}/{self.exp_config['experiment_name']}/best_{self.exp_config['experiment_name']}_episode_{ep_number}.statedict"
+                self.agent.save_model(save_path)
+
+            elif ep_ret > self.best_ret:
+                self.best_ret = ep_ret
+                save_path = f"{self.runner_config['model_save_dir']}/{self.exp_config['experiment_name']}/best_{self.exp_config['experiment_name']}_episode_{ep_number}.statedict"
+                self.agent.save_model(save_path)
```
Comment (kchian) [inline src/runners/sac.py]: There's gotta be a better way of doing this if statement bruh

File: src/runners/sac.py [hunk 3]
```diff
@@ -157,40 +177,43 @@ def eval(self):
...
                 torch.save(self.safety_actor_critic.state_dict(), path_name)
             except:
                 pass
 
-        elif self.cfg["save_freq"] > 0 and (n_eps + 1 % self.cfg["save_freq"] == 0):
-            path_name = f"{self.cfg['model_save_path']}/{self.cfg['experiment_name']}_episode_{n_eps}.statedict"
+        elif self.runner_config["save_every_nth_episode"] > 0 and (
+            n_eps + 1 % self.runner_config["save_every_nth_episode"] == 0
+        ):
+            path_name = f"{self.runner_config['model_save_dir']}/{self.exp_config['experiment_name']}_episode_{n_eps}.statedict"
```
Comment (kchian) [inline src/runners/sac.py:211]: Is it intentional that you changed the path in 2 places and not in others?

File: src/runners/sac.py [hunk 4]
```diff
@@ -157,40 +177,43 @@ def eval(self):
...
+
         self.agent.update_best_pct_complete(info)
 
         return val_ep_rets
 
     ## Sid needs to take care of the actor_critic objects
     def checkpoint_model(self, ep_ret, n_eps):
         if ep_ret > self.best_ret:  # and ep_ret > 100):
-            path_name = f"{self.cfg['model_save_path']}/best_{self.cfg['experiment_name']}_episode_{n_eps}.statedict"
+            path_name = f"{self.runner_config['model_save_dir']}/best_{self.exp_config['experiment_name']}_episode_{n_eps}.statedict"
```
Comment (kchian) [inline src/runners/sac.py]: Is it intentional that you changed the output path in 2 places and not in others?

Comment (tanay-gangey) [reply src/runners/sac.py]: Yeah the code in checkpoint_model is legacy and not being used anywhere

Comment (kchian) [reply src/runners/sac.py]: Can we delete it then?

File: src/runners/sac.py (hunk 2 above)
Comment (tanay-gangey) [reply src/runners/sac.py]: Cleaned it up! Thanks!

**Comment by kchian** [comment]
LGTM I suggest waiting for someone that cares more about rl code to approve tho

**Review by sidharthkathpal — DISMISSED** [review]
Can you put the checkpoint_model function in the run function itself, also can you add the load module as well.

**Comment by tanay-gangey** [comment]
Made the changes to load model and also moved the sacing code to the checkpoint_model function!

**Review by sidharthkathpal — APPROVED** [review]

**Pull request merged**